services/QuillLMS/app/models/teacher_notification.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: teacher_notifications
+#
+#  id                :bigint           not null, primary key
+#  email_sent        :datetime
+#  notification_type :text             not null
+#  params            :jsonb
+#  created_at        :datetime         not null
+#  updated_at        :datetime         not null
+#  user_id           :bigint           not null
+#
+# Indexes
+#
+#  index_teacher_notifications_on_user_id  (user_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (user_id => users.id)
+#
+class TeacherNotification < ApplicationRecord
+  NOTIFICATION_TYPES = [
+    STUDENT_COMPLETED_DIAGNOSTIC = 'student-completed-diganostic',
+    STUDENT_COMPLETED_ALL_DIAGNOSTIC_RECOMMENDATIONS = 'student-completed-all-diagnostic-recommendations',
+    STUDENT_COMPLETED_ALL_ASSIGNED_ACTIVITIES = 'student-completed-all-assigned-activities'
+  ]
+
+  NOTIFICATION_TYPE_VALIDATION = {
+    STUDENT_COMPLETED_DIAGNOSTIC => [
+      :student_name,
+      :classroom_name,
+      :diagnostic_name
+    ],
+    STUDENT_COMPLETED_ALL_DIAGNOSTIC_RECOMMENDATIONS => [
+      :student_name,
+      :classroom_name
+    ],
+    STUDENT_COMPLETED_ALL_ASSIGNED_ACTIVITIES => [
+      :student_name,
+      :classroom_name
+    ]
+  }
+
+  belongs_to :user
+
+  validates :notification_type, presence: true, inclusion: {in: NOTIFICATION_TYPES}
+  validate :params_must_be_hash, :params_payload_has_required_keys
+
+  private def params_is_hash?
+    params.is_a?(Hash)
+  end
+
+  private def params_must_be_hash
+    return if params_is_hash?
+
+    errors.add(:params, "must be a hash")
+  end
+
+  private def params_payload_has_required_keys
+    return unless params_is_hash?
+
+    return if params.keys.map(&:to_s).to_set == NOTIFICATION_TYPE_VALIDATION[notification_type].map(&:to_s).to_set
```

Comment:
Nit: set inclusion would be a bit more flexible than set equality here. We don't care if the payload has more stuff than necessary, but we do care if the payload is lacking a key.

Reply:
@happythenewsad My thought on this was that we'll know exactly what keys we do want, and that additional keys will only be noise.  By doing equality instead of inclusion, my hope was to prevent bloat to these JSON blobs.  Having extra keys probably won't hurt anything other than potentially taking up space in the database, but what are your thoughts on preferring to allow extra keys instead of restricting them?

Reply:
This reasoning makes sense to me - carry on.

I was thinking that it would be annoying to have specs fail when adding a new expected key to the payload, but that's more of a feature than a bug, given that we've suffered from heterogenous jsonb schemata in the past.

Reply:
I ended up doing a refactor based on @dandrabik 's feedback, which has allowed us to avoid a custom validator of this type.  Does this address your concerns here?